overlord/servicestate/quota_control.go

```diff
@@ -0,0 +1,260 @@
+	// validate that the snap exists
+	_, err := snapstate.CurrentInfo(st, name)
+	if err != nil {
+		return err
+	}
+
+	// check that the snap is not already in a group
+	for _, grp := range allGrps {
+		if strutil.ListContains(grp.Snaps, name) {
+			return fmt.Errorf("snap already in quota group %q", grp.Name)
+		}
+	}
+	return nil
+}
+
+// CreateQuota attempts to create the specified quota group with the specified
+// snaps in it.
+// TODO: should this use something like QuotaGroupUpdate with fewer fields?
+func CreateQuota(st *state.State, name string, parentName string, snaps []string, memoryLimit quantity.Size) error {
+	st.Lock()
+	defer st.Unlock()
+
+	// ensure that the quota group does not exist yet
+	allGrps, err := AllQuotas(st)
+	if err != nil {
+		return err
+	}
+
+	if _, ok := allGrps[name]; ok {
+		return fmt.Errorf("group %q already exists", name)
+	}
+
+	// make sure the specified snaps exist and aren't currently in another group
+	for _, sn := range snaps {
+		if err := validateSnapForAddingToGroup(st, sn, allGrps); err != nil {
+			return fmt.Errorf("cannot use snap %q in group %q: %v", sn, name, err)
+		}
+	}
+
+	// make sure that the parent group exists if we are creating a sub-group
+	var grp, parentGrp *quota.Group
+	if parentName != "" {
+		var ok bool
+		parentGrp, ok = allGrps[parentName]
+		if !ok {
+			return fmt.Errorf("cannot create group under non-existent parent group %q", parentName)
+		}
+
+		grp, err = parentGrp.NewSubGroup(name, memoryLimit)
+		if err != nil {
+			return err
+		}
+	} else {
+		// make a new group
+		grp, err = quota.NewGroup(name, memoryLimit)
+		if err != nil {
+			return err
+		}
+	}
+
+	// put the snaps in the group
+	grp.Snaps = snaps
+
+	updatedGrps := []*quota.Group{grp}
+	if parentName != "" {
+		updatedGrps = append(updatedGrps, parentGrp)
+	}
+
+	// update the modified groups in state
+	allGrps, err = PatchQuotasState(st, updatedGrps...)
+	if err != nil {
+		return err
+	}
+
+	// ensure the snap services with the group
+	if err := ensureSnapServicesForGroup(st, grp, allGrps); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// RemoveQuota deletes the specific quota group. Any snaps currently in the
+// quota will no longer be in any quota group, even if the quota group being
+// removed is a sub-group.
+// TODO: currently this only supports removing leaf sub-group groups, it doesn't
+// support removing parent quotas, but probably it makes sense to allow that too
+func RemoveQuota(st *state.State, name string) error {
+	st.Lock()
+	defer st.Unlock()
+
+	allGrps, err := AllQuotas(st)
+	if err != nil {
+		return err
+	}
+
+	// first get the group for later before it is deleted from state
+	grp, ok := allGrps[name]
+	if !ok {
+		return fmt.Errorf("cannot remove non-existent quota group %q", name)
+	}
+
+	// XXX: remove this limitation eventually
+	if len(grp.SubGroups) != 0 {
+		return fmt.Errorf("cannot remove quota group with sub-groups, remove the sub-groups first")
```

Comment:
this needs a test, can be in a follow-up

Reply:
done in https://github.com/snapcore/snapd/pull/10216